account-compression/programs/account-compression/src/macros.rs

```diff
@@ -0,0 +1,110 @@
+        match ($header.get_max_depth(), $header.get_max_buffer_size()) {
+            (3, 8) => _merkle_tree_depth_size_apply_fn!(3, 8, $($arg)*),
+            (5, 8) => _merkle_tree_depth_size_apply_fn!(5, 8, $($arg)*),
+            (14, 64) => _merkle_tree_depth_size_apply_fn!(14, 64, $($arg)*),
+            (14, 256) => _merkle_tree_depth_size_apply_fn!(14, 256, $($arg)*),
+            (14, 1024) => _merkle_tree_depth_size_apply_fn!(14, 1024, $($arg)*),
+            (14, 2048) => _merkle_tree_depth_size_apply_fn!(14, 2048, $($arg)*),
+            (15, 64) => _merkle_tree_depth_size_apply_fn!(15, 64, $($arg)*),
+            (16, 64) => _merkle_tree_depth_size_apply_fn!(16, 64, $($arg)*),
+            (17, 64) => _merkle_tree_depth_size_apply_fn!(17, 64, $($arg)*),
+            (18, 64) => _merkle_tree_depth_size_apply_fn!(18, 64, $($arg)*),
+            (19, 64) => _merkle_tree_depth_size_apply_fn!(19, 64, $($arg)*),
+            (20, 64) => _merkle_tree_depth_size_apply_fn!(20, 64, $($arg)*),
+            (20, 256) => _merkle_tree_depth_size_apply_fn!(20, 256, $($arg)*),
+            (20, 1024) => _merkle_tree_depth_size_apply_fn!(20, 1024, $($arg)*),
+            (20, 2048) => _merkle_tree_depth_size_apply_fn!(20, 2048, $($arg)*),
+            (24, 64) => _merkle_tree_depth_size_apply_fn!(24, 64, $($arg)*),
+            (24, 256) => _merkle_tree_depth_size_apply_fn!(24, 256, $($arg)*),
+            (24, 512) => _merkle_tree_depth_size_apply_fn!(24, 512, $($arg)*),
+            (24, 1024) => _merkle_tree_depth_size_apply_fn!(24, 1024, $($arg)*),
+            (24, 2048) => _merkle_tree_depth_size_apply_fn!(24, 2048, $($arg)*),
+            (26, 512) => _merkle_tree_depth_size_apply_fn!(26, 512, $($arg)*),
+            (26, 1024) => _merkle_tree_depth_size_apply_fn!(26, 1024, $($arg)*),
+            (26, 2048) => _merkle_tree_depth_size_apply_fn!(26, 2048, $($arg)*),
+            (30, 512) => _merkle_tree_depth_size_apply_fn!(30, 512, $($arg)*),
+            (30, 1024) => _merkle_tree_depth_size_apply_fn!(30, 1024, $($arg)*),
+            (30, 2048) => _merkle_tree_depth_size_apply_fn!(30, 2048, $($arg)*),
+            _ => {
+                msg!("Failed to apply {} on concurrent merkle tree with max depth {} and max buffer size {}",
+                    stringify!($func),
+                    $header.get_max_depth(),
+                    $header.get_max_buffer_size()
+                );
+                err!(AccountCompressionError::ConcurrentMerkleTreeConstantsError)
+            }
+        }
+    };
+}
+
+/// This applies a given function on a mutable ConcurrentMerkleTree
+#[macro_export]
+macro_rules! merkle_tree_apply_fn {
+    ($header:ident, $id:ident, $bytes:ident, $func:ident, $($arg:tt)*) => {
+        _merkle_tree_apply_fn!($header, $id, $bytes, $func, TreeLoad::Mutable, $($arg)*)
+    };
+}
+
+/// This applies a given function on a read-only ConcurrentMerkleTree
+#[macro_export]
+macro_rules! merkle_tree_apply_fn_immutable {
```

Comment:
nit: sorry, just realizing now, since it's more idiomatic to specify when something is mutable rather than immutable, so this can be  `merkle_tree_apply_fn` and the other one becomes `merkle_tree_apply_fn_mut`. Not sure if you have outside users depending on this

Reply:
Not sure if I have outside users consuming this merkle tree method either to be honest.

Would it be better to just tag `merkle_tree_apply_fn` as `#[deprecated]` and instead rename to `merkle_tree_apply_fn_mutable` and `merkle_tree_apply_fn_immutable` ?

Reply:
https://crates.io/crates/spl-account-compression/reverse_dependencies -- you can always talk with metaplex? Honestly, not a big deal either way, but if you move to version 0.2, you can break, according to semver

Reply:
I checked their code - no usage of the macros. 

Changed names to match your suggestion. Thanks 🙏